corehq/apps/builds/utils.py

```diff
@@ -40,3 +44,61 @@ def extract_build_info_from_filename(content_disposition):
     else:
         raise ValueError('Could not find filename like {!r} in {!r}'.format(
             pattern, content_disposition))
+
+
+@memoized
+def get_latest_version_at_time(config, target_time):
+    """
+    Get the latest CommCare version that was available at a given time.
+    Excludes superuser-only versions.
+    Menu items are already in chronological order (newest last).
+    If no target time is provided, return the latest version available now.
+
+    Args:
+        config: CommCareBuildConfig instance
+        target_time: datetime or string in ISO format, or None for latest version
+    """
+    if not target_time:
+        return config.get_default().version
+
+    if isinstance(target_time, str):
+        target_time = datetime.strptime(target_time, ISO_DATETIME_FORMAT)
+
+    # Iterate through menu items in reverse (newest to oldest)
+    for item in reversed(config.menu):
+        if item.superuser_only:
+            continue
+        try:
+            build_time = get_build_time(item.build.version)
+            if build_time and build_time <= target_time:
+                return item.build.version
+        except KeyError:
+            continue
+
+    return None
+
+
+@memoized
```

Comment:
The use of `memoized` makes me nervous, as my understanding is that the cached value will stick around until it is explicitly cleared, or the machine restarts. Say we created a version, this ran for that version, and then we had to delete the version. This function would still return the old build time, despite the version no longer existing, right? Worse still, because the cache is local to the machine, one webworker might return the old value, while a different webworker, that hadn't previously processed the request, would return the current value. Does it seem safer to perform manual, per-request caching? I.e. create a dictionary and insert values as the request generates them, but throw that dictionary away at the end of the request.

Reply:
Thank you for noticing it! [I verified in our doc](https://commcare-hq.readthedocs.io/caching_and_memoization.html#lifecycle):
> Memoized on any other function/callable:
The cache lives on the callable, so if it’s globally scoped and never gets garbage collected, neither does the cache


So I'm changing it to a quickcache, and clear it every 10 minutes: 44d9db5

Reply:
Is the intention 10 minutes? `100 * 60` seems to be 100 minutes, right? I'd still prefer a completely local, request-based caching mechanism, because I think it's more in-line with what you're trying to do, but I'll defer to your judgment here. This code works. The downside is that A) caching is harder to debug, B) quickcache is making redis/network requests, and C) we can still run into that inconsistent state due to a stale cache, but now that stale cache will expire after the timeout.

Reply:
Hi Matt, my initial intention is 10 minutes. But now I think 100 minutes is better. I think we will only run into inconsistent state if we release a version then delete it, which I believe this rarely happen. So no need to worry about (A) and (C). By setting `memoize_timeout`, we reduce Redis requests by serving cached values directly from memory if the same process makes repeated calls.

Reply:
I changed my mind!!! Quickcache added a lot of overhead and resulted in worse performance issue, so I implemented a dictionary to store the cache as you suggested: 83aecc62dbf1d35ee75654fc405b673ad7bc1024

Reply:
Love it! Daniel did mention to me, however, that the problem with the initial `@memoized` solution wasn't the caching so much as the _global_ caching. Rather than passing the cache in explicitly, and having to write the logic to insert `None` into the cache when an exception is thrown, for example, it might be cleaner to create an object that can limit the scope of the memoization. But entirely optional -- what you implemented is what I had asked for